pkg/scheduler/framework/v1alpha1/framework.go

```diff
@@ -34,22 +34,23 @@ import (
 // framework is the component responsible for initializing and running scheduler
 // plugins.
 type framework struct {
-	registry                  Registry
-	nodeInfoSnapshot          *schedulernodeinfo.Snapshot
-	waitingPods               *waitingPodsMap
-	pluginNameToWeightMap     map[string]int
-	queueSortPlugins          []QueueSortPlugin
-	preFilterPlugins          []PreFilterPlugin
-	filterPlugins             []FilterPlugin
-	postFilterPlugins         []PostFilterPlugin
-	scorePlugins              []ScorePlugin
-	scoreWithNormalizePlugins []ScoreWithNormalizePlugin
-	reservePlugins            []ReservePlugin
-	preBindPlugins            []PreBindPlugin
-	bindPlugins               []BindPlugin
-	postBindPlugins           []PostBindPlugin
-	unreservePlugins          []UnreservePlugin
-	permitPlugins             []PermitPlugin
+	registry                   Registry
+	nodeInfoSnapshot           *schedulernodeinfo.Snapshot
+	waitingPods                *waitingPodsMap
+	pluginNameToWeightMap      map[string]int
+	queueSortPlugins           []QueueSortPlugin
+	preFilterPlugins           []PreFilterPlugin
+	preFilterWithUpdatePlugins []PreFilterWithUpdatePlugin
```

Comment:
I'm not sure I like the idea of separating the two plugins. In a user's perspective, what they know is just a phase called "PreFilter" which "PreFilterPlugins" should be responsible for. IMO it's better to keep update related logic into another interface.

I'd propose to keep `preFilterPlugins` only, and then add a `MetaUpdater()` method in PreFilterPlugin like this:

```go
type PreFilterPlugin interface {
	Plugin
	PreFilter(pc *PluginContext, p *v1.Pod) *Status
	MetaUpdater() MetaUpdater
}
```

And `MetaUpdater` is an interface with AddPod() and RemovePod():

```go
type MetaUpdater interface {
	AddPod(pc *PluginContext, podToSchedule *v1.Pod, podToAdd *v1.Pod, nodeInfo *schedulernodeinfo.NodeInfo) *Status
	RemovePod(pc *PluginContext, podToSchedule *v1.Pod, podToRemove *v1.Pod, nodeInfo *schedulernodeinfo.NodeInfo) *Status
}
```

So `RunPreFilter[Add|Remove]Pod` method can be updated:

```go
func (f *framework) RunPreFilterAddPod(pc *PluginContext, podToSchedule *v1.Pod,
	podToAdd *v1.Pod, nodeInfo *schedulernodeinfo.NodeInfo) *Status {
	for _, pl := range f.preFilterPlugins {
		status := pl.MetaUpdater().AddPod(pc, podToSchedule, podToAdd, nodeInfo)
```

Reply:
The idea of the separate interfaces is that the extra methods are optional. And then we only iterate once to identify the plugins that implement the interface.
These private slices are not user (developer) facing.
And we have done the same for `scorePlugins` and `scoreWithNormalizePlugins`

Reply:
Right, the idea is to make the methods optional, those slices are implementation details only.

Reply:
Sry to confuse you. I should have commented it on the interface definitions.

So basically, instead of defining two interfaces (`PreFilterPlugin` and `PreFilterWithUpdatePlugin`), I'm leaning towards to keep one interface `PreFilterPlugin`.

With the above extra interface funtion `MetaUpdater()`, implementations that returns nil will get simply bypassed without invoking its `AddPod()` and `RemovePod()` upon preemption. Right now, we're relying on typecasting to tell whether it's a `PreFilterPlugin` or `PreFilterWithUpdatePlugin` implementation, which I'm not sure I like the way.

Reply:
The above suggestion is similar like what we did in GVK:

https://github.com/kubernetes/kubernetes/blob/d62ebe31f0c9bb73217acba847cc4f5f72138823/staging/src/k8s.io/apimachinery/pkg/runtime/schema/interfaces.go#L28

Basically an API object can choose not to omit GVK by implementing `GroupVersionKind()` to simply return nil, so that API machinery won't call further GVK related methods upon this object.

Reply:
this sounds good to me, it will also solve the confusion when a plugin implements only AddPod but not RemovePod for example, which will result in a failure in the type cast, and so AddPod will not be called contrary to what the developer may think.

Reply:
Would it be worth making the optional normalize for score plugins consistent with this suggestion?

Reply:
I think so, but we can have it in another PR.

Reply:
Done. I named the interface "Updater" only, I think it is sufficient.

Reply:
> With the above extra interface funtion `MetaUpdater()`, implementations that returns nil will get simply bypassed without invoking its `AddPod()` and `RemovePod()` upon preemption. Right now, we're relying on typecasting to tell whether it's a `PreFilterPlugin` or `PreFilterWithUpdatePlugin` implementation, which I'm not sure I like the way.

I feel the same way, it's better to implement it as a noop. I sent https://github.com/kubernetes/kubernetes/pull/83042 to remove the scoreWithNormalize interface.